core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -813,7 +813,14 @@ private[spark] class BlockManager(
               false
           }
         } else {
-          memoryStore.putBytes(blockId, size, level.memoryMode, () => bytes)
+          val memoryMode = level.memoryMode
+          memoryStore.putBytes(blockId, size, memoryMode, () => {
+            if (memoryMode == MemoryMode.OFF_HEAP) {
```

Comment:
(Actually, I think we need to check `memoryMode.useOffHeap` here.)

Assume `memoryMode.useOffHeap` is true. We have two cases to consider:

1. `bytes` is on-heap. In this case, we need to copy it into a new direct buffer, and that's what we're doing here.
2. `bytes` is off-heap. In this case, we assume that the caller upstream is managing the memory underlying `bytes`, and `bytes.copy(Platform.allocateDirectBuffer)` becomes a defensive copy. If the caller is not managing this memory, I would call that a bug in the caller's behavior.

In either case, I believe we should be calling `bytes.copy(Platform.allocateDirectBuffer)` when `memoryMode.useOffHeap` is true.

BTW, in my experience tracing this code in the debugger, `bytes` has always been an on-heap buffer.